Search/Factory.php

```diff
@@ -49,13 +53,37 @@ public function makeField($name, $value, $type = Field::TYPE_STRING)
     }
 
     /**
-     * Make a new metadata object representing a mapped
-     * searchable class
+     * Make a new metadata object repesenting a mapped
+     * searchanle class
+     */
+    public function makeClassMetadata($class)
+    {
+        return new ClassMetadata($class);
+    }
+
+    /**
+     * Make a new metadata object representing an index for
+     * a mapped searchable class.
      *
      * @return IndexMetadata
      */
-    public function makeIndexMetadata($class)
+    public function makeIndexMetadata()
+    {
+        return new IndexMetadata();
+    }
+
+    public function makeMetadataField($name)
+    {
+        return new MetadataField($name);
+    }
+
+    public function makeMetadataProperty($path)
```

Comment:
This is a service that enables the user to implement their own classes, for example, the Metadata.

It is not very common, but is required if you want flexibility instead of having classes instantiate their own objects, and it also protects BC - for example in the Sulu Metadata driver we need to instantiate new fields, if we did that manually it would mean that we could not move, rename or change the constructor arguments of the MassiveSearch classes.